4.x typehints 5

ADmad opened the pull request:
Updated `TestSuite` to `View`.

ADmad committed:
Update typehints for TestSuite/

ADmad committed:
Update type hints for Utility/

ADmad committed:
Update psalm's baseline file.

ADmad committed:
Update type hints for Validation/

dereuromark commented on src/Utility/Hash.php:
```diff
@@ -209,11 +209,11 @@ protected static function _splitConditions(string $token): array
      * Check a key against a token.
      *
-     * @param string $key The key in the array being searched.
+     * @param mixed $key The key in the array being searched.
```
We should avoid mixed by using string|int etc

ADmad replied:
Well you can't since `Hash` class deals with arrays where keys don't always have to be string.

dereuromark replied:
True, but usually scalar. As such better to be explicit. Right now you would also allow objects or other arrays etc, thats bad.

ADmad replied:
Specifying a list of all scalar type vs mixed isn't going to provide any practical benefit. Won't make any difference to tools and humans can already know it takes valid array key types by reading description.

dereuromark replied:
Why? You then know that it cannot be any random object for example. As such the benefit is 10x more than mixed already by design.

dereuromark commented on src/Utility/Hash.php:
```diff
@@ -296,7 +296,7 @@ protected static function _matches($data, string $selector): bool
      * @param array $data The data to insert into.
      * @param string $path The path to insert at.
-     * @param array|null $values The values to insert.
+     * @param mixed $values The values to insert.
```
Can we keep this more specific one?

ADmad replied:
No since you can insert any type of value to particular key.

dereuromark replied:
Isnt this iterating inside, so it must be an array of things - saying mixed means you could also work on random objects like Response class, I doubt that is possible or intended.
We usually annotate the intended target input, not all remotely possible ones you could pass without crashing it IMO.

ADmad replied:
Fixing incorrect type was more valuable than listing the possible one. You are welcome to check the code / test case and list the possible types instead.

dereuromark commented on src/Validation/RulesProvider.php:
```diff
@@ -59,9 +59,9 @@ public function __construct($class = Validation::class)
      * @param string $method the validation method to call
      * @param array $arguments the list of arguments to pass to the method
-     * @return bool whether or not the validation rule passed
+     * @return mixed whether or not the validation rule passed
```
You should check/revert all mixed changes.

ADmad replied:
All mixed changes are based on actual code. Adding typehints as per old docblocks causes tests to fail.

dereuromark replied:
What does a method return that so far was boolean? See comment beside it.

ADmad replied:
This is for the magic `__call()` which proxies calls to other to other classes so we really can't guarantee that the return type will always be boolean.

dereuromark replied:
No one can ever guarantee that, nor did we ever do that.
We always have been annotating the expected input and output types, that is what we should continue to do, not all possible wrong ones on top. Otherwise documentation here really doesnt help much, neither for humans nor for tooling.

dereuromark replied:
To clarify: The follow up code will blow up if the documented expectations are not met. That is then outside of this documentation here. 
We could, if we want, also add a bool typehint to assert that on code level of course.

ADmad replied:
Same comment as above. I fixed the incorrect type, you can update it for accuracy.

ADmad committed:
Update typehints for View/

ADmad committed:
Update psalm's baseline file.

ADmad commented:
With this PR I have done a pass for all packages. 

The updates for holy trinity of Database, Datasource and ORM in #13192 still remains. I'll get back to it after taking a breather as trying to property resolve types in those packages leads to quite a bit of hair pulling.

dereuromark commented:
I am strongly against adding more mixed typehints where it is not really true.
We need to be more precise here, as mixed kills basically all introspection, which in those cases above is not really true. Those are not random value storage, but concrete input to operate on.

ADmad commented:
@dereuromark For the 3 instances where the type is changed to `mixed` I have stated the reason. Each change is logically justified and as per actual test cases.

dereuromark commented:
I disagree. Each instance is different, but none of them justify going to mixed which basically means anything. That is not true in those cases. Ideally, you break it down more specifically here.

ADmad commented:
Changing those to `mixed` makes the docblock correct. You can update them to be more accurate.

ADmad commented:
Appveyour build success didn't last even for a day :laughing: 
https://ci.appveyor.com/project/cakephp/cakephp/builds/24355793#L24929

ADmad committed:
Use a custom validation provider who's method returns boolean.

This avoid confusion and ensure return value for RulesProvider::__call() is as documented.
